typing/shape_reduce.ml

```diff
@@ -197,8 +198,11 @@ end) = struct
           | NStruct (items) ->
               begin match Item.Map.find item items with
               | exception Not_found -> nored ()
-              | nf -> force nf |> reset_uid_if_new_binding
+              | nf -> force env nf |> reset_uid_if_new_binding
               end
+          | NLeaf ->
+              (* Happens when projecting from a first-class module *)
+              approx_nf (nored ())
```

Comment:
Do we want a similar case in the `App(f, args)` block above, if `f.desc` is a `NLeaf`? (Can you test this?)

Reply:
That's not clear. I added a case that would be impacted by such a change in [d682f39](https://github.com/ocaml/ocaml/pull/13001/commits/d682f39950ff7d06711d0ef14603010d0bbd0556). 

If in `module FArg = X.F (Arg)`, `X` is actually an instance of a first class module, so the shape of `FArg` is indeed unsolvable. But when reducing the shape of `Farg` "for uid", we don't really care about that since we want the uid of the binding we made when writing `FArg = ... `. We do care however if we reduce the shape of `Farg.t`, which should return `approximated no uid`  (but right now returns `no uid` without the approximated mention.).

Reply:
I would find it more natural to have the same behavior for reduction of functor application and structure projection. If you don't have a strong preference, I would do the same on both sides.

Reply:
I pushed in [b9ba1b9](https://github.com/ocaml/ocaml/pull/13001/commits/b9ba1b99643f6c8ac301d3c9e112cae059eeba9f) a test for `Farg.t` and added a case to effectively mark it as approximated instead of returning an `Internal_error_no_uid`.

I think that maybe `approximated` is not a good name or the best concept: it actually only occurs when accessing items inside first-class-modules ? (But I won't have time to clean this now, I will be away from this Friday until the beginning of April.)

Reply:
Hm, the new code does not look like what I would expect: I don't think the `NApp _` case here is correct. If you don't want to work on this now, I would suggest removing this part of the code from the present PR (it is not really in scope anyway).

More details: projecting out of a NLeaf is the sign of a problem, because NLeaf should not be the shape of a structure. On the other hand, NApp can naturally occur on open terms -- there is an external compilation unit under it. What I had in mind was to add a `NLeaf` case in the `NApp (f, args)` case above, that is, where we currently have `| _ ->` line 190. Just like you did in the projection case, in the case where this catch-all is actually a leaf we know that an approximation took place.

Reply:
> I would find it more natural to have the same behavior for reduction of functor application and structure projection. If you don't have a strong preference, I would do the same on both sides.

I knew you would say that :-) 

But if I change the behavior for application it will break occurrences of modules that are new bindings for the result of functor applications that are approximated. For example, the indexed occurrence of the `FArg` module would be: 

```diff
- Resolved: Index_aliases.26 :
+ Approximated: Index_aliases.26 :
  FArg (File "index_aliases.ml", line 48, characters 5-9)
 ```
 
But in that case `Index_aliases.26` is really the identifier of the definition of `FArg`, it's not an approximation.

If we want to keep symmetry until we find a better solution then I suggest we remove the other case: having some uids registered with an `Internal_error_missing_uid` instead of an `Approximated None` does not have an impact on the actual useful occurrences while having changes such as the one in the diff for uid `Index_aliases.26` will prevent some useful occurrences to be gathered.

If that's ok with you I will rollback the changes i made to this PR that are not related to its initial purpose: drastic performance improvement in the presence of aliases. The approximation issues are not critical right now and we can tackle them when we will have more time.

Reply:
So or messages interleaved twice, I think we agree on rewinding back to the original intent of this PR, I will do that.